include/elf/arc-reloc.def

```diff
@@ -497,3 +497,24 @@ ARC_RELOC_HOWTO(ARC_NPS_CMEM16, 78, \
                 replace_bits16, \
                 dont, \
                 ( ME ( S + A )))
+
+ARC_RELOC_HOWTO(ARC_JLI_SECTOFF, 63, \
+                2, \
+                10, \
+                replace_jli, \
+                bitfield, \
+                ( ( S - JLI ) ))
+
+ARC_RELOC_HOWTO(ARC_JLI_SECTOFF_SIMM12, 79, \
+                2, \
+                12, \
+                replace_simm12_20, \
+                bitfield, \
+                ( ( ( S - JLI ) >> 2 ) ))
+
+ARC_RELOC_HOWTO(ARC_JLI_32, 80, \
```

Comment:
ARC_JLI_SECTOFF_SIMM12 was for the add2 instruction. Here, the address of the function in the JLI table is taken as the offset into the JLI section ((S - JLI) >> 2) (the shift is because of add2) plus the base of the JLI section _when this code is executed_. If you don't move the JLI table it equates to S. If you do move the JLI table, it equates to (JLI_BASE + (S - JLI)) where JLI_BASE is the contents of the AUX register and not the original location in the ELF file.

ARC_JLI_32 is a bit more tricky. When you initialize a static/global variable with the address of a function in the JLI table, the linker must then solve for this address. In this case, the linker can't solve (JLI_BASE + (S - JLI)) but it can solve (JLI + (S - JLI)) or S. That means it's up to the program loader to make sure this is properly changed if the JLI table was moved. For a main application you may want to simply set the jli_base register to "JLI" but if you are loading in multiple targets, you may not want to alter the jli_base AUX register and simply relocate all function pointers to ensure they point into the pre-existing JLI table.